python/datasources/bjs_incarceration.py

```diff
@@ -0,0 +1,321 @@
+        df.loc[~df[std_col.AGE_COL].isin(NON_NULL_RAW_COUNT_GROUPS),
+               [RAW_COL]] = np.nan
+        df = df.drop(columns=[std_col.POPULATION_COL])
+
+    else:
+        df = df.drop(columns=[std_col.POPULATION_COL, RAW_COL])
+
+    return df
+
+
+class BJSIncarcerationData(DataSource):
+
+    @ staticmethod
+    def get_id():
+        return 'BJS_INCARCERATION_DATA'
+
+    @ staticmethod
+    def get_table_name():
+        return 'bjs_incarceration_data'
+
+    def upload_to_gcs(self, _, **attrs):
+        raise NotImplementedError(
+            'upload_to_gcs should not be called for BJSIncarcerationData')
+
+    def write_to_bq(self, dataset, gcs_bucket, **attrs):
+        """
+        Main function for this data source that fetches external data and runs
+        needed cleaning and standardization, and then writes needed tables to
+        BigQuery
+
+        """
+        loaded_tables = load_tables(BJS_PRISONERS_ZIP)
+        df_2 = standardize_table_2_df(loaded_tables[TABLE_2])
+        df_10 = standardize_table_10_df(loaded_tables[TABLE_10])
+        df_13 = standardize_table_13_df(loaded_tables[TABLE_13])
+        df_23 = standardize_table_23_df(loaded_tables[TABLE_23])
+        df_app_2 = standardize_appendix_table_2_df(
+            loaded_tables[APPENDIX_TABLE_2])
+
+        # BJS tables needed per breakdown
+        table_lookup = {
+            f'{std_col.AGE_COL}_{NATIONAL_LEVEL}': [df_10, df_13],
+            f'{std_col.AGE_COL}_{STATE_LEVEL}': [df_2, df_13, df_23],
+            f'{std_col.RACE_OR_HISPANIC_COL}_{NATIONAL_LEVEL}': [df_app_2, df_23],
+            f'{std_col.SEX_COL}_{NATIONAL_LEVEL}': [df_2, df_23],
+            f'{std_col.RACE_OR_HISPANIC_COL}_{STATE_LEVEL}': [df_app_2, df_23],
+            f'{std_col.SEX_COL}_{STATE_LEVEL}': [df_2, df_23],
+        }
+
+        for geo_level in [NATIONAL_LEVEL, STATE_LEVEL]:
+            for breakdown in [std_col.AGE_COL, std_col.RACE_OR_HISPANIC_COL, std_col.SEX_COL]:
+                table_name = f'{breakdown}_{geo_level}'
+
+                df = self.generate_breakdown_df(
+                    breakdown, geo_level, table_lookup[table_name])
+
+                # set / add BQ types
+                column_types = {c: 'STRING' for c in df.columns}
+                for col in BJS_DATA_TYPES:
+                    column_types[std_col.generate_column_name(
+                        col, std_col.PER_100K_SUFFIX)] = 'FLOAT'
+                    column_types[std_col.generate_column_name(
+                        col, std_col.PCT_SHARE_SUFFIX)] = 'FLOAT'
+                column_types[std_col.POPULATION_PCT_COL] = 'FLOAT'
+                if std_col.RACE_INCLUDES_HISPANIC_COL in df.columns:
+                    column_types[std_col.RACE_INCLUDES_HISPANIC_COL] = 'BOOL'
+
+                gcs_to_bq_util.add_df_to_bq(
+                    df, dataset, table_name, column_types=column_types)
+
+    def generate_breakdown_df(self, breakdown, geo_level, table_list, ):
```

Comment:
did you leave the last comma in by accident?

Reply:
oh ya; thanks. I had originally sent a lot more arguments in and then simplified it.